### PR TITLE
#7109 Initial commit.

### PR DESCRIPTION
**Pull request by dragospodariu96**
#7109

### CODE REVIEW

File: backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5565000_sys_gh7109AddUpdateOrderAndInvoiceIdProcess.sql [hunk 1]
```diff
@@ -0,0 +1,85 @@
+-- 2020-08-11T09:03:36.811Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+INSERT INTO AD_Process (AccessLevel,AD_Client_ID,AD_Org_ID,AD_Process_ID,AllowProcessReRun,Classname,CopyFromProcess,Created,CreatedBy,EntityType,IsActive,IsApplySecuritySettings,IsBetaFunctionality,IsDirectPrint,IsNotifyUserAfterExecution,IsOneInstanceOnly,IsReport,IsServerProcess,IsTranslateExcelHeaders,IsUseBPartnerLanguage,LockWaitTimeout,Name,RefreshAllAfterExecution,ShowHelp,Type,Updated,UpdatedBy,Value) VALUES ('3',0,0,584736,'Y','de.metas.banking.process.C_Payment_UpdateOrderAndInvoiceId','N',TO_TIMESTAMP('2020-08-11 12:03:36','YYYY-MM-DD HH24:MI:SS'),100,'D','Y','N','N','N','N','N','N','N','Y','Y',0,'updateOrderAndInvoiceID','Y','N','Java',TO_TIMESTAMP('2020-08-11 12:03:36','YYYY-MM-DD HH24:MI:SS'),100,'updateOrderAndInvoiceID')
```
Comment (metas-ts) [inline backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5565000_sys_gh7109AddUpdateOrderAndInvoiceIdProcess.sql:3]: make sure the process has an understandable name and description

File: backend/de.metas.adempiere.adempiere/base/src/main/java-gen/org/compiere/model/I_C_Order.java [hunk 2]
```diff
@@ -8,6 +8,10 @@
 public interface I_C_Order 
 {
 
+	String COLUMNNAME_AD_InputDataSource_ID = "AD_InputDataSource_ID";
```
Comment (metas-ts) [inline backend/de.metas.adempiere.adempiere/base/src/main/java-gen/org/compiere/model/I_C_Order.java:11]: that's OK, but then please remove this stuff from `de.metas.order.model.I_C_Order`

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceDAO.java [hunk 3]
```diff
@@ -57,6 +58,10 @@
 
 	void save(org.compiere.model.I_C_InvoiceLine invoiceLine);
 
+	Map<OrderId, InvoiceId> getInvoiceIdsForOrderIds(final List<OrderId> orderIds);
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceDAO.java:61]:
```suggestion
	Map<OrderId, InvoiceId> getInvoiceIdsForOrderIds(List<OrderId> orderIds);
```

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceDAO.java [hunk 4]
```diff
@@ -57,6 +58,10 @@
...
+
+	I_C_Invoice getByOrderId(final OrderId orderId);
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceDAO.java:63]:
```suggestion
	I_C_Invoice getByOrderId(OrderId orderId);
```

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceDAO.java [hunk 5]
```diff
@@ -53,6 +62,33 @@ public I_C_Invoice createInvoice(String trxName)
...
 	}
 
+	@Override
+	public Map<OrderId, InvoiceId> getInvoiceIdsForOrderIds(final List<OrderId> orderIds)
+	{
+		Map<OrderId, InvoiceId> orderIdInvoiceIdMap = new HashMap<OrderId, InvoiceId>();
+
+		for (OrderId orderId : orderIds)
+		{
+			orderIdInvoiceIdMap.put(orderId, InvoiceId.ofRepoId(getByOrderId(orderId).getC_Order_ID()));
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceDAO.java:72]: i think there is no point loading them one-by-one.
you can use IQueryBuilder.addInArrayFilter()

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceDAO.java [hunk 6]
```diff
@@ -53,6 +62,33 @@ public I_C_Invoice createInvoice(String trxName)
...
+		}
+		return orderIdInvoiceIdMap;
+	}
+
+	@Override
+	public I_C_Invoice getByOrderId(@NonNull final OrderId orderId)
+	{
+		final I_C_Invoice invoice = Services.get(IQueryBL.class)
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceDAO.java:80]: i think the method is never used. can be removed or made private

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceDAO.java [hunk 7]
```diff
@@ -53,6 +62,33 @@ public I_C_Invoice createInvoice(String trxName)
 		return InterfaceWrapperHelper.create(Env.getCtx(), I_C_Invoice.class, trxName);
 	}
 
+	@Override
+	public Map<OrderId, InvoiceId> getInvoiceIdsForOrderIds(final List<OrderId> orderIds)
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceDAO.java:66]: move this method into AbstractInvoiceDAO. Is has no hardcoded SQL-stuff, so we don't need a dedicated `PlainInvoiceDAO.java` implementation

File: backend/de.metas.business/src/main/java/de/metas/order/impl/AbstractOrderDAO.java [hunk 8]
```diff
@@ -73,6 +77,34 @@ public I_C_Order getById(@NonNull final OrderId orderId)
 		return order;
 	}
 
+	@Override
+	public I_C_Order getByExternalId(@NonNull final ExternalId externalId)
+	{
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/order/impl/AbstractOrderDAO.java:82]: the same comments as in `IInvoiceDAO`

File: backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java [hunk 9]
```diff
@@ -113,6 +114,12 @@ public I_C_Order getById(@NonNull final OrderId orderId)
 		return Services.get(IOrderDAO.class).getById(orderId);
 	}
 
+	@Override
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java:117]: why this method in OrderBL?

File: backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java [hunk 10]
```diff
@@ -419,6 +429,66 @@ public boolean isMatchInvoice(final I_C_Payment payment, final I_C_Invoice invoi
 		return false;
 	}
 
+	@Override
+	public String isPaypalOrCreditCardPayment(final I_C_Payment payment, final int paypalDataSourceId, final int creditCardDataSourceId)
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java:433]: this method and it's invocation in `de.metas.banking.process.C_Payment_UpdateOrderAndInvoiceId#doIt` is unneccesary complex. 
Also:
* if it starts with `is` it should return a boolean
* don't return a String constant, but (*if* this method is not removed), an enum

File: backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5564980_sys_gh7109PaymentRuleColumnAddedToCPaymentTable.sql [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+-- 2020-08-11T07:33:42.404Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+/* DDL */  select update_Column_Translation_From_AD_Element(1143) 
+;
+
+-- 2020-08-11T07:37:32.107Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+UPDATE AD_Column SET DefaultValue='@IsAllocated@=Y',Updated=TO_TIMESTAMP('2020-08-11 10:37:32','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Column_ID=571098
+;
```
[the remark this answers is not in the document]
Comment (dragospodariu96) [reply backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5564980_sys_gh7109PaymentRuleColumnAddedToCPaymentTable.sql:19]: This is removed a few lines down.

File: backend/de.metas.business/src/main/java/de/metas/order/impl/AbstractOrderDAO.java (hunk 8 above)
Comment (dragospodariu96) [reply backend/de.metas.business/src/main/java/de/metas/order/impl/AbstractOrderDAO.java:81]: made it private

File: backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java (hunk 9 above)
Comment (dragospodariu96) [reply backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java:117]: not used, removed method

File: backend/de.metas.business/src/main/java/de/metas/payment/interceptor/C_Payment.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+import org.adempiere.ad.modelvalidator.annotations.Interceptor;
+import org.adempiere.ad.modelvalidator.annotations.ModelChange;
+import org.compiere.model.I_C_Order;
+import org.compiere.model.I_C_Payment;
+import org.compiere.model.ModelValidator;
+import org.springframework.stereotype.Component;
+
+@Interceptor(I_C_Payment.class)
+@Component
+public class C_Payment
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/payment/interceptor/C_Payment.java:44]: (i wrote this comment after the two that follow in this class)

I understand that this is aiming at completed payments that get their respective invoice from your process.
* I think that your process is the only one to set C_Order_ID and C_Invoice_ID on an already completed payment
* for not-yet-completed payment, we AFAIU have Cristian's code

=> so, rather that have this model interceptor that "incidentally" allocated the payment,
make the allocation explicitly in the PaymentBL-code that also sets the C_Order_ID and C_Invoice_ID

File: backend/de.metas.business/src/main/java/de/metas/payment/interceptor/PaymentModuleInterceptor.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+	protected void registerInterceptors(final IModelValidationEngine engine)
+	{
+		engine.addModelValidator(C_Payment.INSTANCE); // FRESH-348
+
+	}
+
+	@Override
+	protected void registerCallouts(final IProgramaticCalloutProvider calloutsRegistry)
+	{
+		calloutsRegistry.registerAnnotatedCallout(C_Payment.INSTANCE); // FRESH-348
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/payment/interceptor/PaymentModuleInterceptor.java:47]: * C_Payment has no callout methods AFAIU
* Generally, think we should ditch the whole intercheptor; see my comment at C_Payment

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5565000_sys_gh7109AddUpdateOrderAndInvoiceIdProcess.sql [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+-- 2020-08-25T07:34:01.397Z
+-- URL zum Konzept
+UPDATE AD_Process_Trl SET IsTranslated='Y', Name='Set Order and Invoice',Updated=TO_TIMESTAMP('2020-08-25 10:34:01','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Language='en_US' AND AD_Process_ID=584736
+;
+
+-- 2020-08-25T07:34:47.225Z
+-- URL zum Konzept
+UPDATE AD_Process_Trl SET Name='Bestellung und Rechnung festlegen',Updated=TO_TIMESTAMP('2020-08-25 10:34:47','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Language='nl_NL' AND AD_Process_ID=584736
+;
```
[the remark this answers is not in the document]
Comment (dragospodariu96) [reply backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5565000_sys_gh7109AddUpdateOrderAndInvoiceIdProcess.sql:144]: Added description and updated name

**Review by metas-ts — APPROVED** [review]

**Pull request merged**